client/paper/whisker.js

```diff
@@ -0,0 +1,171 @@
+      this.ctx = this.canvas.getContext('2d');
+    }
+
+    this.whiskers.push(whisker);
+
+    return whisker;
+  }
+
+  async destroyWhisker(whisker) {
+    this.whiskers = this.whiskers.filter(other => other !== whisker);
+
+    if (this.whiskers.length === 0) {
+      clearInterval(this.update);
+    }
+  }
+
+  async update() {
+    const papers = await this.workerContext.paper.get('papers');
+
+    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
+    this.whiskers.forEach(whisker => whisker.update(papers, this.ctx));
+    this.ctx.commit();
+  }
+}
+
+class Whisker extends EventEmitter {
+  constructor({
+    whiskerFactory,
+    paperNumber,
+    direction = 'up',
+    whiskerLength = 0.7,
+    requiredData = [],
+    color = 'rgb(255, 0, 0)',
+  }) {
+    super();
+
+    this.paperNumber = paperNumber;
+    this.direction = direction;
+    this.whiskerLength = whiskerLength;
+    this.requiredData = requiredData;
+    this.color = color;
+
+    this._pointAtData = null;
+    this._lastWhiskerEnd = null;
+    this._whiskerFactory = whiskerFactory;
+  }
+
+  update(papers, ctx) {
+    const points = papers[this.paperNumber].points;
+
+    let segment = [points.topLeft, points.topRight];
+    if (this.direction === 'right') segment = [points.topRight, points.bottomRight];
+    if (this.direction === 'down') segment = [points.bottomRight, points.bottomLeft];
+    if (this.direction === 'left') segment = [points.bottomLeft, points.topLeft];
+
+    const whiskerStart = {
+      x: (segment[0].x + segment[1].x) / 2,
+      y: (segment[0].y + segment[1].y) / 2,
+    };
+    const whiskerEnd = {
+      x: whiskerStart.x + (segment[1].y - segment[0].y) * this.whiskerLength,
+      y: whiskerStart.y - (segment[1].x - segment[0].x) * this.whiskerLength,
+    };
+
+    if (
+      !this._pointAtData ||
+      !papers[this._pointAtData.paperNumber] ||
+      // Try keeping `pointAtData` stable if possible.
+      !this._intersectsPaper(whiskerStart, whiskerEnd, papers[this._pointAtData.paperNumber])
+    ) {
+      let newPointAtData = null;
+      Object.keys(papers).forEach(otherPaperNumber => {
+        if (otherPaperNumber === this.paperNumber) return;
+        if (this._intersectsPaper(whiskerStart, whiskerEnd, papers[otherPaperNumber])) {
+          newPointAtData = { paperNumber: otherPaperNumber, paper: papers[otherPaperNumber] };
+        }
+      });
+
+      if (this._pointAtData) {
+        this.emit('paperRemoved', this._pointAtData);
+      }
+
+      if (newPointAtData) {
+        this.emit('paperAdded', newPointAtData);
+      }
+
+      this._pointAtData = newPointAtData;
+    }
+
+    // render whisker with dot animation
+    ctx.clearRect(0, 0, ctx.width, ctx.height);
+    ctx.fillStyle = ctx.strokeStyle = this.color;
+    ctx.beginPath();
+    ctx.moveTo(whiskerStart.x, whiskerStart.y);
+    ctx.lineTo(whiskerEnd.x, whiskerEnd.y);
+    ctx.stroke();
+
+    const dotFraction = Math.abs(Math.sin(Date.now() / 600));
```

Comment:
I've changed the dot animation to alternate direction. I wanted to avoid implying a direction of the dataflow which depends on what the whisker does and instead just visualize connectedness.